File: GJOH6/src/task2/service/JsonOperation.java
package task2.service;

import task2.model.Contact;

import java.util.List;

/**
 * The interface contains methods for working with the file
 */
public interface JsonOperation {

    /**
     * Method saves contacts list to the .json file
     *
     * @param filename filename
     * @param contacts List of contacts
     */
    void saveToJsonFile(String filename, List<Contact> contacts);

    /**
     * Method reads contacts from .json file
     *
     * @param filename filename
     * @return List of Contacts
     */
    List<Contact> readFromJsonFile(String filename);
}

File: GJOH7/service/json_operation.py
import json
from typing import List
from model.phone_contact import PhoneContact as Contact


class JsonOperation:
    """
    The class contains methods for working with a json file
    """
    def __init__(self):
        self.contacts = []

    def read_from_json_file(self, filename: str):
        """
        Method reads contacts from json file
        :param filename filename 
        """
        self.contacts = []
        try:
            with open(filename, 'r') as f:
                data = json.load(f)
                for item in data:
                    contact = Contact(
                        item['firstname'], item['lastname'], item.get('numbers', []))
                    self.add_contact(contact)
                return self.contacts
        except IOError:
            print("Error opening or reading json file")
            return []

    def save_to_json_file(self, contacts: List[Contact], filename: str):
        """
        Method saves contacts list to the json file
        :param contacts List of contacts
        :param filename String filename
        """
        try:
            with open(filename, 'w') as f:
                data = []
                for contact in contacts:
                    data.append({
                        'firstname': contact.firstname,
                        'lastname': contact.lastname,
                        'numbers': contact.numbers,
                    })
                json.dump(data, f)
        except IOError:
            print("Error opening or writing to json file")

    def add_contact(self, contact):
        self.contacts.append(contact)
